src/algorithm/rotate.rs

```diff
@@ -0,0 +1,251 @@
+use num_traits::{Float, FromPrimitive};
+use types::{Point, Polygon, LineString};
+use algorithm::centroid::Centroid;
+
+// rotate a slice of points "angle" degrees about an origin
+// origin can be an arbitrary point, pass &Point::new(0., 0.)
+// for the actual origin
+fn rotation_matrix<T>(angle: T, origin: &Point<T>, points: &[Point<T>]) -> Vec<Point<T>>
+    where T: Float
+{
+    let cos_theta = angle.to_radians().cos();
+    let sin_theta = angle.to_radians().sin();
+    let x0 = origin.x();
+    let y0 = origin.y();
+    points
+        .iter()
+        .map(|point| {
+                 let x = point.x() - x0;
+                 let y = point.y() - y0;
+                 Point::new(x * cos_theta - y * sin_theta + x0,
+                            x * sin_theta + y * cos_theta + y0)
+             })
+        .collect::<Vec<_>>()
+}
+
+pub trait Rotate<T> {
+    /// Rotate a Geometry around its centroid by an angle, in degrees
+    ///
+    /// Positive angles are counter-clockwise, and negative angles are clockwise rotations.
+    ///
+    /// ```
+    /// use geo::{Point, LineString};
+    /// use geo::algorithm::rotate::{Rotate};
+    ///
+    /// let mut vec = Vec::new();
+    /// vec.push(Point::new(0.0, 0.0));
+    /// vec.push(Point::new(5.0, 5.0));
+    /// vec.push(Point::new(10.0, 10.0));
+    /// let linestring = LineString(vec);
+    /// let rotated = linestring.rotate(-45.0);
+    /// let mut correct = Vec::new();
+    /// correct.push(Point::new(-2.0710678118654755, 5.0));
+    /// correct.push(Point::new(5.0, 5.0));
+    /// correct.push(Point::new(12.071067811865476, 5.0));
+    /// let correct_ls = LineString(correct);
+    /// assert_eq!(rotated, correct_ls);
+    /// ```
+    fn rotate(&self, angle: T) -> Self where T: Float;
+}
```

Comment:
I think you should be able to make a generic implementation for this, right? Something like:

```rust
impl<T> Rotate<T> for G<T>
    where T: Float,
          G: Centroid, RotatePoint
{
    fn rotate(&self, angle: T) -> Self {
        self.rotate_point(self.centroid())
    }
}
```

I haven't actually tested this though

Reply:
@urschrei I'm revisiting this. Regarding polygon rotation, what's wrong with using the centroid of a complex polygon (using the weighted center) instead of how it works right now (just using the centroid of the exterior)?

Reply:
IIRC it's because that's what GEOS does; even if the polygon is complex, it's treated as simple when calculating the centroid. I don't know _why_ this is, though.